File: dotnet/tests/Obscura.Cdp.Tests/ExecutionContextOwnership.cs
using System.Text.Json.Nodes;

using Xunit;

namespace Obscura.Cdp.Tests;

/// <summary>
/// The xUnit port of <c>crates/obscura-cdp/tests/execution_context_ownership.rs</c>.
/// </summary>
[Collection(CdpDomainCollection.Name)]
public sealed class ExecutionContextOwnership
{
    private static async Task<(string Target, string Session)> CreateAndAttachAsync(
        CdpContext ctx,
        string url,
        ulong id)
    {
        CdpResponse created = await CoreCdp.DispatchAsync(
            ctx, id, "Target.createTarget", new JsonObject { ["url"] = url }, null);
        Assert.True(created.Error is null, $"createTarget failed: {created.Error?.Message}");
        string target = created.Result!["targetId"].AsString()!;
        CdpResponse attached = await CoreCdp.DispatchAsync(
            ctx,
            id + 1,
            "Target.attachToTarget",
            new JsonObject { ["targetId"] = target, ["flatten"] = true },
            null);
        return (target, attached.Result!["sessionId"].AsString()!);
    }

    private static (long Id, string UniqueId, string Origin) LatestDefaultContext(
        CdpContext ctx,
        string session)
    {
        CdpEvent found = ctx.PendingEvents
            .Where(e => e.Method == "Runtime.executionContextCreated"
                && e.SessionId == session
                && e.Params.Get("context").Get("auxData").Get("isDefault").AsBool() == true)
            .LastOrDefault()
            ?? throw new InvalidOperationException("missing default context");
        JsonNode? context = found.Params.Get("context");
        return (
            context.Get("id").AsI64()!.Value,
            context.Get("uniqueId").AsString()!,
            context.Get("origin").AsString()!);
    }

    [Fact]
    public async Task NonblankCreateTargetExposesOnlyTheCommittedDocumentContext()
    {
        var ctx = CdpContext.New();
        using IDisposable owned = CoreCdp.Owned(ctx);
        (_, string session) = await CreateAndAttachAsync(
            ctx, "data:text/html,<title>committed</title>", 1);
        ctx.PendingEvents.Clear();
        await CoreCdp.DispatchAsync(ctx, 3, "Runtime.enable", new JsonObject(), session);

        (long id, _, string origin) = LatestDefaultContext(ctx, session);
        Assert.Equal(1, id);
        Assert.StartsWith("data:text/html,", origin, StringComparison.Ordinal);
        Assert.DoesNotContain(
            ctx.PendingEvents,
            e => e.Method == "Runtime.executionContextCreated"
                && e.Params.Get("context").Get("origin").AsString() == "about:blank");
    }

    [Fact]
    public async Task DefaultContextIdentityIsPageOwnedForIdAndUniqueId()
    {
        var ctx = CdpContext.New();
        using IDisposable owned2 = CoreCdp.Owned(ctx);
        (_, string first) = await CreateAndAttachAsync(ctx, "about:blank", 1);
        (_, string second) = await CreateAndAttachAsync(ctx, "about:blank", 10);
        await CoreCdp.DispatchAsync(ctx, 20, "Runtime.enable", new JsonObject(), first);
        (long id, string uniqueId, _) = LatestDefaultContext(ctx, first);

        CdpResponse foreignId = await CoreCdp.DispatchAsync(
            ctx,
            21,
            "Runtime.evaluate",
            new JsonObject { ["expression"] = "1", ["contextId"] = id },
            second);
        Assert.Contains("Cannot find context", foreignId.Error!.Message, StringComparison.Ordinal);

        CdpResponse foreignUnique = await CoreCdp.DispatchAsync(
            ctx,
            22,
            "Runtime.evaluate",
            new JsonObject { ["expression"] = "1", ["uniqueContextId"] = uniqueId },
            second);
        Assert.Contains("Cannot find context", foreignUnique.Error!.Message, StringComparison.Ordinal);

        CdpResponse foreignCall = await CoreCdp.DispatchAsync(
            ctx,
            23,
            "Runtime.callFunctionOn",
            new JsonObject
            {
                ["functionDeclaration"] = "() => 1",
                ["executionContextId"] = id,
                ["returnByValue"] = true,
            },
            second);
        Assert.Contains("Cannot find context", foreignCall.Error!.Message, StringComparison.Ordinal);

        await CoreCdp.DispatchAsync(
            ctx,
            24,
            "Page.navigate",
            new JsonObject
            {
                ["url"] = "data:text/html,<p>replacement</p>",
                ["waitUntil"] = "load",
            },
            first);

        (ulong RequestId, string Method, JsonObject Params)[] stalePairs =
        [
            (25UL, "Runtime.evaluate",
                new JsonObject { ["expression"] = "1", ["uniqueContextId"] = uniqueId }),
            (26UL, "Runtime.callFunctionOn", new JsonObject
            {
                ["functionDeclaration"] = "() => 1",
                ["executionContextId"] = id,
                ["returnByValue"] = true,
            }),
        ];
        foreach ((ulong requestId, string method, JsonObject parameters) in stalePairs)
        {
            CdpResponse stale = await CoreCdp.DispatchAsync(ctx, requestId, method, parameters, first);
            Assert.NotNull(stale.Error);
            Assert.Contains("Cannot find context", stale.Error!.Message, StringComparison.Ordinal);
        }
    }

    [Fact]
    public async Task NavigatingOnePagePreservesTheOtherPagesIsolatedContext()
    {
        var ctx = CdpContext.New();
        using IDisposable owned3 = CoreCdp.Owned(ctx);
        (_, string first) = await CreateAndAttachAsync(ctx, "about:blank", 1);
        (_, string second) = await CreateAndAttachAsync(ctx, "about:blank", 10);
        foreach ((ulong id, string session) in new[] { (20UL, first), (21UL, second) })
        {
            await CoreCdp.DispatchAsync(ctx, id, "Runtime.enable", new JsonObject(), session);
        }

        long firstIsolated = (await CoreCdp.CdpAsync(
            ctx,
            30,
            "Page.createIsolatedWorld",
            new JsonObject { ["worldName"] = "first-utility" },
            first))["executionContextId"].AsI64()!.Value;
        long secondIsolated = (await CoreCdp.CdpAsync(
            ctx,
            31,
            "Page.createIsolatedWorld",
            new JsonObject { ["worldName"] = "second-utility" },
            second))["executionContextId"].AsI64()!.Value;

        await CoreCdp.DispatchAsync(
            ctx,
            32,
            "Page.navigate",
            new JsonObject
            {
                ["url"] = "data:text/html,<p>first replacement</p>",
                ["waitUntil"] = "load",
            },
            first);

        CdpResponse secondStillRoutes = await CoreCdp.DispatchAsync(
            ctx,
            33,
            "Runtime.evaluate",
            new JsonObject { ["expression"] = "1", ["contextId"] = secondIsolated },
            second);
        Assert.Null(secondStillRoutes.Error);

        CdpResponse firstIsStale = await CoreCdp.DispatchAsync(
            ctx,
            34,
            "Runtime.evaluate",
            new JsonObject { ["expression"] = "1", ["contextId"] = firstIsolated },
            first);
        Assert.NotNull(firstIsStale.Error);
        Assert.Contains("Cannot find context", firstIsStale.Error!.Message, StringComparison.Ordinal);
    }

    [Fact]
    public async Task AttachedIsolatedContextIdsShareTheCurrentPageGlobalForNow()
    {
        var ctx = CdpContext.New();
        using IDisposable owned4 = CoreCdp.Owned(ctx);
        (_, string session) = await CreateAndAttachAsync(ctx, "about:blank", 1);
        await CoreCdp.DispatchAsync(
            ctx,
            3,
            "Runtime.evaluate",
            new JsonObject { ["expression"] = "globalThis.realmProbe = 'default'" },
            session);
        long isolated = (await CoreCdp.CdpAsync(
            ctx,
            4,
            "Page.createIsolatedWorld",
            new JsonObject { ["worldName"] = "bookkeeping-only" },
            session))["executionContextId"].AsI64()!.Value;

        CdpResponse isolatedResponse = await CoreCdp.DispatchAsync(
            ctx,
            5,
            "Runtime.evaluate",
            new JsonObject
            {
                ["expression"] =
                    "(function(){ globalThis.realmProbe += '-isolated'; return globalThis.realmProbe; })()",
                ["contextId"] = isolated,
                ["returnByValue"] = true,
            },
            session);
        Assert.True(
            isolatedResponse.Error is null,
            $"isolated evaluate failed: {isolatedResponse.Error?.Message}");
        Assert.Equal(
            "default-isolated", isolatedResponse.Result!["result"]!["value"].AsString());

        JsonNode defaultValue = await CoreCdp.CdpAsync(
            ctx,
            6,
            "Runtime.evaluate",
            new JsonObject { ["expression"] = "globalThis.realmProbe", ["returnByValue"] = true },
            session);
        Assert.Equal("default-isolated", defaultValue["result"]!["value"].AsString());
    }

    [Fact]
    public async Task NavigationContextEventsPreserveOrderForEveryRuntimeAttachment()
    {
        var ctx = CdpContext.New();
        using IDisposable owned5 = CoreCdp.Owned(ctx);
        (string target, string first) = await CreateAndAttachAsync(ctx, "about:blank", 1);
        CdpResponse attached = await CoreCdp.DispatchAsync(
            ctx,
            3,
            "Target.attachToTarget",
            new JsonObject { ["targetId"] = target, ["flatten"] = true },
            null);
        string second = attached.Result!["sessionId"].AsString()!;
        foreach ((ulong id, string session) in new[] { (10UL, first), (11UL, second) })
        {
            await CoreCdp.DispatchAsync(ctx, id, "Runtime.enable", new JsonObject(), session);
        }

        await CoreCdp.DispatchAsync(
            ctx,
            12,
            "Page.createIsolatedWorld",
            new JsonObject { ["worldName"] = "utility" },
            first);
        ctx.PendingEvents.Clear();

        await CoreCdp.DispatchAsync(
            ctx,
            13,
            "Page.navigate",
            new JsonObject
            {
                ["url"] = "data:text/html,<p>replacement</p>",
                ["waitUntil"] = "load",
            },
            first);

        List<string> firstMethods = [.. ctx.PendingEvents
            .Where(e => e.SessionId == first
                && (e.Method == "Page.lifecycleEvent"
                    || e.Method == "Page.frameNavigated"
                    || e.Method.StartsWith("Runtime.executionContext", StringComparison.Ordinal)))
            .Select(e => e.Method == "Page.lifecycleEvent"
                ? $"{e.Method}:{e.Params.Get("name").AsStringOr(string.Empty)}"
                : e.Method)];
        Assert.Equal(
            [
                "Page.lifecycleEvent:init",
                "Runtime.executionContextsCleared",
                "Page.frameNavigated",
                "Runtime.executionContextCreated",
            ],
            firstMethods.Take(4));
        Assert.Equal(
            2,
            firstMethods.Count(m => string.Equals(m, "Runtime.executionContextCreated", StringComparison.Ordinal)));

        List<string> secondMethods = [.. ctx.PendingEvents
            .Where(e => e.SessionId == second
                && e.Method.StartsWith("Runtime.executionContext", StringComparison.Ordinal))
            .Select(e => e.Method)];
        Assert.Equal(
            [
                "Runtime.executionContextsCleared",
                "Runtime.executionContextCreated",
                "Runtime.executionContextCreated",
            ],
            secondMethods);
    }

    [Fact]
    public async Task RuntimeAndBindingEventsUseTheOwningPagesDefaultContext()
    {
        var ctx = CdpContext.New();
        using IDisposable owned6 = CoreCdp.Owned(ctx);
        (_, string first) = await CreateAndAttachAsync(ctx, "about:blank", 1);
        (_, string second) = await CreateAndAttachAsync(ctx, "about:blank", 10);
        foreach ((ulong id, string session) in new[] { (20UL, first), (21UL, second) })
        {
            await CoreCdp.DispatchAsync(ctx, id, "Runtime.enable", new JsonObject(), session);
        }

        (long firstContext, _, _) = LatestDefaultContext(ctx, first);
        await CoreCdp.DispatchAsync(
            ctx, 22, "Runtime.addBinding", new JsonObject { ["name"] = "probe" }, second);
        ctx.PendingEvents.Clear();

        await CoreCdp.DispatchAsync(
            ctx,
            23,
            "Page.navigate",
            new JsonObject
            {
                ["url"] = "data:text/html,<script>console.log('owned');probe('bound')</script>",
                ["waitUntil"] = "load",
            },
            second);
        (long secondContext, _, _) = LatestDefaultContext(ctx, second);
        Assert.NotEqual(firstContext, secondContext);

        CdpEvent console = ctx.PendingEvents.First(e =>
            e.Method == "Runtime.consoleAPICalled" && e.SessionId == second);
        Assert.Equal(secondContext, console.Params.Get("executionContextId").AsI64());
        CdpEvent binding = ctx.PendingEvents.First(e =>
            e.Method == "Runtime.bindingCalled" && e.SessionId == second);
        Assert.Equal(secondContext, binding.Params.Get("executionContextId").AsI64());
    }
}

File: .reference/obscura/crates/obscura-cdp/tests/execution_context_ownership.rs
use obscura_cdp::dispatch::{dispatch, CdpContext};
use obscura_cdp::types::CdpRequest;
use serde_json::{json, Value};

async fn cdp(
    ctx: &mut CdpContext,
    id: u64,
    method: &str,
    params: Value,
    session_id: Option<&str>,
) -> obscura_cdp::types::CdpResponse {
    dispatch(&CdpRequest {
        id,
        method: method.to_string(),
        params,
        session_id: session_id.map(str::to_string),
    }, ctx).await
}

async fn create_and_attach(ctx: &mut CdpContext, url: &str, id: u64) -> (String, String) {
    let created = cdp(ctx, id, "Target.createTarget", json!({"url": url}), None).await;
    assert!(created.error.is_none(), "createTarget failed: {:?}", created.error);
    let target = created.result.unwrap()["targetId"].as_str().unwrap().to_string();
    let attached = cdp(
        ctx, id + 1, "Target.attachToTarget",
        json!({"targetId": target, "flatten": true}), None,
    ).await;
    let session = attached.result.unwrap()["sessionId"].as_str().unwrap().to_string();
    (target, session)
}

fn latest_default_context(ctx: &CdpContext, session: &str) -> (i64, String, String) {
    let context = &ctx.pending_events.iter().rev().find(|event| {
        event.method == "Runtime.executionContextCreated"
            && event.session_id.as_deref() == Some(session)
            && event.params["context"]["auxData"]["isDefault"] == true
    }).expect("missing default context").params["context"];
    (
        context["id"].as_i64().unwrap(),
        context["uniqueId"].as_str().unwrap().to_string(),
        context["origin"].as_str().unwrap().to_string(),
    )
}

#[tokio::test(flavor = "current_thread")]
async fn nonblank_create_target_exposes_only_the_committed_document_context() {
    let mut ctx = CdpContext::new();
    let (_, session) = create_and_attach(
        &mut ctx, "data:text/html,<title>committed</title>", 1,
    ).await;
    ctx.pending_events.clear();
    cdp(&mut ctx, 3, "Runtime.enable", json!({}), Some(&session)).await;

    let (id, _, origin) = latest_default_context(&ctx, &session);
    assert_eq!(id, 1);
    assert!(origin.starts_with("data:text/html,"));
    assert!(!ctx.pending_events.iter().any(|event| {
        event.method == "Runtime.executionContextCreated"
            && event.params["context"]["origin"] == "about:blank"
    }));
}

#[tokio::test(flavor = "current_thread")]
async fn default_context_identity_is_page_owned_for_id_and_unique_id() {
    let mut ctx = CdpContext::new();
    let (_, first) = create_and_attach(&mut ctx, "about:blank", 1).await;
    let (_, second) = create_and_attach(&mut ctx, "about:blank", 10).await;
    cdp(&mut ctx, 20, "Runtime.enable", json!({}), Some(&first)).await;
    let (id, unique_id, _) = latest_default_context(&ctx, &first);

    let foreign_id = cdp(
        &mut ctx, 21, "Runtime.evaluate",
        json!({"expression": "1", "contextId": id}), Some(&second),
    ).await;
    assert!(foreign_id.error.unwrap().message.contains("Cannot find context"));
    let foreign_unique = cdp(
        &mut ctx, 22, "Runtime.evaluate",
        json!({"expression": "1", "uniqueContextId": unique_id.clone()}), Some(&second),
    ).await;
    assert!(foreign_unique.error.unwrap().message.contains("Cannot find context"));

    let foreign_call = cdp(
        &mut ctx, 23, "Runtime.callFunctionOn",
        json!({
            "functionDeclaration": "() => 1",
            "executionContextId": id,
            "returnByValue": true,
        }),
        Some(&second),
    ).await;
    assert!(foreign_call.error.unwrap().message.contains("Cannot find context"));

    cdp(
        &mut ctx, 24, "Page.navigate",
        json!({"url": "data:text/html,<p>replacement</p>", "waitUntil": "load"}),
        Some(&first),
    ).await;
    for (request_id, params) in [
        (25, json!({"expression": "1", "uniqueContextId": unique_id})),
        (26, json!({
            "functionDeclaration": "() => 1",
            "executionContextId": id,
            "returnByValue": true,
        })),
    ] {
        let method = if request_id == 25 { "Runtime.evaluate" } else { "Runtime.callFunctionOn" };
        let stale = cdp(&mut ctx, request_id, method, params, Some(&first)).await;
        assert!(stale.error.unwrap().message.contains("Cannot find context"));
    }
}

#[tokio::test(flavor = "current_thread")]
async fn navigating_one_page_preserves_the_other_pages_isolated_context() {
    let mut ctx = CdpContext::new();
    let (_, first) = create_and_attach(&mut ctx, "about:blank", 1).await;
    let (_, second) = create_and_attach(&mut ctx, "about:blank", 10).await;
    for (id, session) in [(20, &first), (21, &second)] {
        cdp(&mut ctx, id, "Runtime.enable", json!({}), Some(session)).await;
    }
    let first_isolated = cdp(
        &mut ctx, 30, "Page.createIsolatedWorld",
        json!({"worldName": "first-utility"}), Some(&first),
    ).await.result.unwrap()["executionContextId"].as_i64().unwrap();
    let second_isolated = cdp(
        &mut ctx, 31, "Page.createIsolatedWorld",
        json!({"worldName": "second-utility"}), Some(&second),
    ).await.result.unwrap()["executionContextId"].as_i64().unwrap();

    cdp(
        &mut ctx, 32, "Page.navigate",
        json!({"url": "data:text/html,<p>first replacement</p>", "waitUntil": "load"}),
        Some(&first),
    ).await;

    let second_still_routes = cdp(
        &mut ctx, 33, "Runtime.evaluate",
        json!({"expression": "1", "contextId": second_isolated}), Some(&second),
    ).await;
    assert!(second_still_routes.error.is_none());
    let first_is_stale = cdp(
        &mut ctx, 34, "Runtime.evaluate",
        json!({"expression": "1", "contextId": first_isolated}), Some(&first),
    ).await;
    assert!(first_is_stale.error.unwrap().message.contains("Cannot find context"));
}

#[tokio::test(flavor = "current_thread")]
async fn attached_isolated_context_ids_share_the_current_page_global_for_now() {
    let mut ctx = CdpContext::new();
    let (_, session) = create_and_attach(&mut ctx, "about:blank", 1).await;
    cdp(
        &mut ctx, 3, "Runtime.evaluate",
        json!({"expression": "globalThis.realmProbe = 'default'"}), Some(&session),
    ).await;
    let isolated = cdp(
        &mut ctx, 4, "Page.createIsolatedWorld",
        json!({"worldName": "bookkeeping-only"}), Some(&session),
    ).await.result.unwrap()["executionContextId"].as_i64().unwrap();
    let isolated_response = cdp(
        &mut ctx, 5, "Runtime.evaluate",
        json!({
            "expression": "(function(){ globalThis.realmProbe += '-isolated'; return globalThis.realmProbe; })()",
            "contextId": isolated,
            "returnByValue": true,
        }),
        Some(&session),
    ).await;
    assert!(isolated_response.error.is_none(), "isolated evaluate failed: {:?}", isolated_response.error);
    let isolated_value = isolated_response.result.unwrap();
    assert_eq!(isolated_value["result"]["value"], "default-isolated");
    let default_value = cdp(
        &mut ctx, 6, "Runtime.evaluate",
        json!({"expression": "globalThis.realmProbe", "returnByValue": true}),
        Some(&session),
    ).await.result.unwrap();
    assert_eq!(default_value["result"]["value"], "default-isolated");
}

#[tokio::test(flavor = "current_thread")]
async fn navigation_context_events_preserve_order_for_every_runtime_attachment() {
    let mut ctx = CdpContext::new();
    let (target, first) = create_and_attach(&mut ctx, "about:blank", 1).await;
    let attached = cdp(
        &mut ctx, 3, "Target.attachToTarget",
        json!({"targetId": target, "flatten": true}), None,
    ).await;
    let second = attached.result.unwrap()["sessionId"].as_str().unwrap().to_string();
    for (id, session) in [(10, &first), (11, &second)] {
        cdp(&mut ctx, id, "Runtime.enable", json!({}), Some(session)).await;
    }
    cdp(
        &mut ctx, 12, "Page.createIsolatedWorld",
        json!({"worldName": "utility"}), Some(&first),
    ).await;
    ctx.pending_events.clear();

    cdp(
        &mut ctx, 13, "Page.navigate",
        json!({"url": "data:text/html,<p>replacement</p>", "waitUntil": "load"}),
        Some(&first),
    ).await;

    let first_methods = ctx.pending_events.iter().filter(|event| {
        event.session_id.as_deref() == Some(first.as_str())
            && (event.method == "Page.lifecycleEvent"
                || event.method == "Page.frameNavigated"
                || event.method.starts_with("Runtime.executionContext"))
    }).map(|event| {
        if event.method == "Page.lifecycleEvent" {
            format!("{}:{}", event.method, event.params["name"].as_str().unwrap_or(""))
        } else {
            event.method.clone()
        }
    }).collect::<Vec<_>>();
    assert_eq!(&first_methods[..4], [
        "Page.lifecycleEvent:init",
        "Runtime.executionContextsCleared",
        "Page.frameNavigated",
        "Runtime.executionContextCreated",
    ]);
    assert_eq!(first_methods.iter().filter(|method| {
        method.as_str() == "Runtime.executionContextCreated"
    }).count(), 2);

    let second_methods = ctx.pending_events.iter().filter(|event| {
        event.session_id.as_deref() == Some(second.as_str())
            && event.method.starts_with("Runtime.executionContext")
    }).map(|event| event.method.as_str()).collect::<Vec<_>>();
    assert_eq!(second_methods, [
        "Runtime.executionContextsCleared",
        "Runtime.executionContextCreated",
        "Runtime.executionContextCreated",
    ]);
}

#[tokio::test(flavor = "current_thread")]
async fn runtime_and_binding_events_use_the_owning_pages_default_context() {
    let mut ctx = CdpContext::new();
    let (_, first) = create_and_attach(&mut ctx, "about:blank", 1).await;
    let (_, second) = create_and_attach(&mut ctx, "about:blank", 10).await;
    for (id, session) in [(20, &first), (21, &second)] {
        cdp(&mut ctx, id, "Runtime.enable", json!({}), Some(session)).await;
    }
    let (first_context, _, _) = latest_default_context(&ctx, &first);
    cdp(
        &mut ctx,
        22,
        "Runtime.addBinding",
        json!({"name": "probe"}),
        Some(&second),
    ).await;
    ctx.pending_events.clear();

    cdp(
        &mut ctx,
        23,
        "Page.navigate",
        json!({
            "url": "data:text/html,<script>console.log('owned');probe('bound')</script>",
            "waitUntil": "load",
        }),
        Some(&second),
    ).await;
    let (second_context, _, _) = latest_default_context(&ctx, &second);
    assert_ne!(first_context, second_context);

    let console = ctx.pending_events.iter().find(|event| {
        event.method == "Runtime.consoleAPICalled"
            && event.session_id.as_deref() == Some(second.as_str())
    }).expect("missing console event");
    assert_eq!(console.params["executionContextId"], second_context);
    let binding = ctx.pending_events.iter().find(|event| {
        event.method == "Runtime.bindingCalled"
            && event.session_id.as_deref() == Some(second.as_str())
    }).expect("missing binding event");
    assert_eq!(binding.params["executionContextId"], second_context);
}
